File: internal/dsl/loader.go
package dsl

import (
	"fmt"
	"sync"
)

// Loader DSL 加载器
type Loader struct {
	parser    *Parser
	validator *Validator
	schema    *OntologySchema
	mu        sync.RWMutex
}

// NewLoader 创建新的加载器
func NewLoader(filePath string) *Loader {
	return &Loader{
		parser: NewParser(filePath),
	}
}

// Load 加载并验证 Schema
func (l *Loader) Load() error {
	l.mu.Lock()
	defer l.mu.Unlock()

	schema, err := l.parser.Parse()
	if err != nil {
		return fmt.Errorf("failed to parse schema: %w", err)
	}

	validator := NewValidator(schema)
	if err := validator.Validate(); err != nil {
		return fmt.Errorf("schema validation failed: %w", err)
	}

	l.schema = schema
	l.validator = validator

	return nil
}

// GetSchema 获取当前 Schema
func (l *Loader) GetSchema() *OntologySchema {
	l.mu.RLock()
	defer l.mu.RUnlock()
	return l.schema
}

// Reload 重新加载 Schema（用于热重载）
func (l *Loader) Reload() error {
	return l.Load()
}

// GetObjectType 根据名称获取对象类型
func (l *Loader) GetObjectType(name string) (*ObjectType, error) {
	schema := l.GetSchema()
	if schema == nil {
		return nil, fmt.Errorf("schema not loaded")
	}

	for i := range schema.ObjectTypes {
		if schema.ObjectTypes[i].Name == name {
			return &schema.ObjectTypes[i], nil
		}
	}

	return nil, fmt.Errorf("object type '%s' not found", name)
}

// ListObjectTypes 列出所有对象类型
func (l *Loader) ListObjectTypes() []ObjectType {
	schema := l.GetSchema()
	if schema == nil {
		return []ObjectType{}
	}

	result := make([]ObjectType, len(schema.ObjectTypes))
	copy(result, schema.ObjectTypes)
	return result
}

// GetLinkType 根据名称获取关系类型
func (l *Loader) GetLinkType(name string) (*LinkType, error) {
	schema := l.GetSchema()
	if schema == nil {
		return nil, fmt.Errorf("schema not loaded")
	}

	for i := range schema.LinkTypes {
		if schema.LinkTypes[i].Name == name {
			return &schema.LinkTypes[i], nil
		}
	}

	return nil, fmt.Errorf("link type '%s' not found", name)
}

// ListLinkTypes 列出所有关系类型
func (l *Loader) ListLinkTypes() []LinkType {
	schema := l.GetSchema()
	if schema == nil {
		return []LinkType{}
	}

	result := make([]LinkType, len(schema.LinkTypes))
	copy(result, schema.LinkTypes)
	return result
}

// GetOutgoingLinks 获取对象的出边关系
func (l *Loader) GetOutgoingLinks(objectTypeName string) []LinkType {
	schema := l.GetSchema()
	if schema == nil {
		return []LinkType{}
	}

	var result []LinkType
	for _, lt := range schema.LinkTypes {
		if lt.SourceType == objectTypeName {
			result = append(result, lt)
		}
	}
	return result
}

// GetIncomingLinks 获取对象的入边关系
func (l *Loader) GetIncomingLinks(objectTypeName string) []LinkType {
	schema := l.GetSchema()
	if schema == nil {
		return []LinkType{}
	}

	var result []LinkType
	for _, lt := range schema.LinkTypes {
		if lt.TargetType == objectTypeName {
			result = append(result, lt)
		}
	}
	return result
}

File: src/main/java/com/mypalantir/meta/Loader.java
package com.mypalantir.meta;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class Loader {
    private final Parser parser;
    private final String systemSchemaPath;
    private final ReadWriteLock lock = new ReentrantReadWriteLock();
    private OntologySchema schema;

    public Loader(String filePath) {
        this(filePath, null);
    }

    public Loader(String filePath, String systemSchemaPath) {
        this.parser = new Parser(filePath);
        this.systemSchemaPath = systemSchemaPath;
    }

    public void load() throws IOException, Validator.ValidationException {
        lock.writeLock().lock();
        try {
            OntologySchema parsedSchema = parser.parse();
            
            // 如果指定了系统 schema 路径，加载并合并
            if (systemSchemaPath != null && !systemSchemaPath.isEmpty()) {
                Parser systemParser = new Parser(systemSchemaPath);
                OntologySchema systemSchema = systemParser.parse();
                
                // 合并两个 schema
                parsedSchema = mergeSchemas(systemSchema, parsedSchema);
            }
            
            Validator schemaValidator = new Validator(parsedSchema);
            schemaValidator.validate();
            this.schema = parsedSchema;
        } finally {
            lock.writeLock().unlock();
        }
    }

    /**
     * 合并系统 schema 和用户 schema
     * 系统 schema 的对象和关系会优先，用户 schema 中的同名对象和关系会被忽略
     */
    private OntologySchema mergeSchemas(OntologySchema systemSchema, OntologySchema userSchema) {
        OntologySchema merged = new OntologySchema();
        
        // 使用用户 schema 的版本和命名空间（如果用户 schema 有的话）
        merged.setVersion(userSchema.getVersion() != null ? userSchema.getVersion() : systemSchema.getVersion());
        merged.setNamespace(userSchema.getNamespace() != null ? userSchema.getNamespace() : systemSchema.getNamespace());
        
        // 合并对象类型：系统 schema 优先，然后添加用户 schema 中不存在的对象类型
        List<ObjectType> mergedObjectTypes = new ArrayList<>();
        if (systemSchema.getObjectTypes() != null) {
            mergedObjectTypes.addAll(systemSchema.getObjectTypes());
        }
        if (userSchema.getObjectTypes() != null) {
            for (ObjectType userOt : userSchema.getObjectTypes()) {
                // 检查是否已存在同名对象类型
                boolean exists = mergedObjectTypes.stream()
                    .anyMatch(ot -> ot.getName().equals(userOt.getName()));
                if (!exists) {
                    mergedObjectTypes.add(userOt);
                }
            }
        }
        merged.setObjectTypes(mergedObjectTypes);
        
        // 合并关系类型：系统 schema 优先，然后添加用户 schema 中不存在的关系类型
        List<LinkType> mergedLinkTypes = new ArrayList<>();
        if (systemSchema.getLinkTypes() != null) {
            mergedLinkTypes.addAll(systemSchema.getLinkTypes());
        }
        if (userSchema.getLinkTypes() != null) {
            for (LinkType userLt : userSchema.getLinkTypes()) {
                // 检查是否已存在同名关系类型
                boolean exists = mergedLinkTypes.stream()
                    .anyMatch(lt -> lt.getName().equals(userLt.getName()));
                if (!exists) {
                    mergedLinkTypes.add(userLt);
                }
            }
        }
        merged.setLinkTypes(mergedLinkTypes);
        
        return merged;
    }

    public OntologySchema getSchema() {
        lock.readLock().lock();
        try {
            return schema;
        } finally {
            lock.readLock().unlock();
        }
    }

    public void reload() throws IOException, Validator.ValidationException {
        load();
    }

    public ObjectType getObjectType(String name) throws NotFoundException {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null || currentSchema.getObjectTypes() == null) {
            throw new NotFoundException("schema not loaded");
        }

        for (ObjectType ot : currentSchema.getObjectTypes()) {
            if (name.equals(ot.getName())) {
                return ot;
            }
        }

        throw new NotFoundException("object type '" + name + "' not found");
    }

    public List<ObjectType> listObjectTypes() {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null) {
            return List.of();
        }
        return currentSchema.getObjectTypes() != null ? List.copyOf(currentSchema.getObjectTypes()) : List.of();
    }

    public LinkType getLinkType(String name) throws NotFoundException {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null || currentSchema.getLinkTypes() == null) {
            throw new NotFoundException("schema not loaded");
        }

        for (LinkType lt : currentSchema.getLinkTypes()) {
            if (name.equals(lt.getName())) {
                return lt;
            }
        }

        throw new NotFoundException("link type '" + name + "' not found");
    }

    public List<LinkType> listLinkTypes() {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null) {
            return List.of();
        }
        return currentSchema.getLinkTypes() != null ? List.copyOf(currentSchema.getLinkTypes()) : List.of();
    }

    public List<LinkType> getOutgoingLinks(String objectTypeName) {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null || currentSchema.getLinkTypes() == null) {
            return List.of();
        }

        return currentSchema.getLinkTypes().stream()
            .filter(lt -> objectTypeName.equals(lt.getSourceType()))
            .toList();
    }

    public List<LinkType> getIncomingLinks(String objectTypeName) {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null || currentSchema.getLinkTypes() == null) {
            return List.of();
        }

        return currentSchema.getLinkTypes().stream()
            .filter(lt -> objectTypeName.equals(lt.getTargetType()))
            .toList();
    }

    public List<DataSourceConfig> listDataSources() {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null) {
            return List.of();
        }
        return currentSchema.getDataSources() != null ? List.copyOf(currentSchema.getDataSources()) : List.of();
    }

    public DataSourceConfig getDataSourceById(String id) throws NotFoundException {
        OntologySchema currentSchema = getSchema();
        if (currentSchema == null || currentSchema.getDataSources() == null) {
            throw new NotFoundException("schema not loaded");
        }

        DataSourceConfig dataSource = currentSchema.getDataSourceById(id);
        if (dataSource == null) {
            throw new NotFoundException("data source '" + id + "' not found");
        }
        return dataSource;
    }

    public static class NotFoundException extends Exception {
        public NotFoundException(String message) {
            super(message);
        }
    }
}
